README.md
# WWW-Crawler-Mojo

WWW::Crawler::Mojo is a web crawling framework written in Perl on top of mojo toolkit, allowing you to write your own crawler rapidly. 

***This software is considered to be alpha quality and isn't recommended for regular usage.***

## Features

* Easy to rule your crawler.
* Allows to use [Mojo::URL] for URL manipulations, [Mojo::Message::Response] for response manipulation and [Mojo::DOM] for DOM inspection.
* Internally uses [Mojo::UserAgent] which is a full featured non-blocking I/O HTTP and WebSocket user agent, with IPv6, TLS, SNI, IDNA, HTTP/SOCKS5 proxy, Comet (long polling), keep-alive, connection pooling, timeout, cookie, multipart, gzip compression and multiple event loop.
* Throttle the connection with max connection and max connection per host options.
* Depth detection.
* Tracks 301 HTTP redirects.
* Detects network errors and retry with your own rules.
* Shuffles queue periodically if indicated.
* Crawls beyond basic authentication.
* Crawls even error documents.
* Form submitting emulation.

[Mojo::URL]:http://mojolicio.us/perldoc/Mojo/URL
[Mojo::DOM]:http://mojolicio.us/perldoc/Mojo/DOM
[Mojo::Message::Response]:http://mojolicio.us/perldoc/Mojo/Message/Response
[Mojo::UserAgent]:http://mojolicio.us/perldoc/Mojo/UserAgent

## Requirements

* Perl 5.16 or higher
* Mojolicious 6.0 or higher

## Installation

    $ curl -L cpanmin.us | perl - -n WWW::Crawler::Mojo

## Synopsis

    use WWW::Crawler::Mojo;
    
    my $bot = WWW::Crawler::Mojo->new;
    
    $bot->on(res => sub {
        my ($bot, $scrape, $job, $res) = @_;
        
        for my $job ($scrape->($css_selector)) {
        	if (...) {
            	$bot->enqueue($job);
            }
        }
    });
    
    $bot->enqueue('http://example.com/');
    $bot->crawl;

## Documentation

* [WWW::Crawler::Mojo](http://search.cpan.org/perldoc?WWW%3A%3ACrawler%3A%3AMojo)
* [WWW::Crawler::Mojo::Job](http://search.cpan.org/perldoc?WWW%3A%3ACrawler%3A%3AMojo%3A%3AJob)
* [WWW::Crawler::Mojo::UserAgent](http://search.cpan.org/perldoc?WWW%3A%3ACrawler%3A%3AMojo%3A%3AUserAgent)
* [WWW::Crawler::Mojo::Queue](http://search.cpan.org/perldoc?WWW%3A%3ACrawler%3A%3AMojo%3A%3AQueue)
* [WWW::Crawler::Mojo::Queue::Memory](http://search.cpan.org/perldoc?WWW%3A%3ACrawler%3A%3AMojo%3A%3AQueue%3A%3AMemory)
* [WWW::Crawler::Mojo::ScraperUtil](http://search.cpan.org/perldoc?WWW%3A%3ACrawler%3A%3AMojo%3A%3AScraperUtil)

## Examples

Restricting scraping URLs by status code.

    $bot->on(res => sub {
        my ($bot, $scrape, $job, $res) = @_;
        return unless ($res->code == 200);
        $bot->enqueue($_) for $scrape->();
    });

Restricting scraping URLs by host.

    $bot->on(res => sub {
        my ($bot, $scrape, $job, $res) = @_;
        return unless if ($job->url->host eq 'example.com');
        $bot->enqueue($_) for $scrape->();
    });

Restrict following URLs by depth.

    $bot->on(res => sub {
        my ($bot, $scrape, $job, $res) = @_;
        
        for my $job ($scrape->()) {
            next unless ($job->depth < 5)
            $bot->enqueue($job);
        }
    });

Restrict following URLs by host.

    $bot->on(res => sub {
        my ($bot, $scrape, $job, $res) = @_;
        
        for my $job ($scrape->()) {
            $bot->enqueue($job) if $job->url->host eq 'example.com';
        }
    });

Restrict following URLs by referrer's host.

    $bot->on(res => sub {
        my ($bot, $scrape, $job, $res) = @_;
        
        for my $job ($scrape->()) {
            $bot->enqueue($job) if $job->referrer->url->host eq 'example.com';
        }
    });

Excepting following URLs by path.

    $bot->on(res => sub {
        my ($bot, $scrape, $job, $res) = @_;
        
        for my $job ($scrape->()) {
            $bot->enqueue($job) unless ($job->url->path =~ qr{^/foo/});
        }
    });

Crawl only preset URLs.

    $bot->on(res => sub {
        my ($bot, $scrape, $job, $res) = @_;
        # DO SOMETHING
    });
	
	$bot->enqueue(
    	'http://example.com/1',
    	'http://example.com/3',
    	'http://example.com/5',
    );
	
	$bot->crawl;

Speed up.

    $bot->max_conn(5);
    $bot->max_conn_per_host(5);

Authentication. The user agent automatically reuses the credential for the host.

    $bot->enqueue('http://jamadam:password@example.com');

You can fulfill any prerequisites such as login form submittion so that a login session will be established with cookie or something which you don't have to worry about.

    my $bot = WWW::Crawler::Mojo->new;
    $bot->ua->post('http://example.com/admin/login', form => {
        username => 'jamadam',
        password => 'password',
    });
    $bot->enqueue('http://example.com/admin/');
    $bot->crawl

## Other examples

* [WWW-Flatten](https://github.com/jamadam/WWW-Flatten)
* See the scripts under the example directory.

## Broad crawling

Although the module is only well tested for "focused crawl" at this point,
you can also use it for endless crawling by taking special care of memory usage including;

* Restrict queue size by yourself.
* Replace redundant detecter code.

	$bot->queue->redundancy(sub {...});

## Copyright

Copyright (C) jamadam

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.



Changes
Revision history for WWW-Crawler-Mojo

0.27 2021/03/17

0.26 2021/03/17
  - Fix tests for latest mojo changes.

0.25 2019/11/13
  - Reduce memory usage.

0.24 2019/08/16
  - Fixed a bug where wrong schemas in links would not be omitted.

0.23 2019/04/02
  - Now enqueue methods returns the actual jobs
  - Now it's breadth-first on memory capacity control

0.22 2019/03/27
  - Fixed default error callback

0.21 2019/03/06
  - Adjusted to latest Mojolicious release.
  - Fix a bug where url including white spaces haven't been treated propery.
  - Form submition now includes passwords and dates.
  - Started Appveyor to test against windows.

0.20 2017/01/17
  - Added an ability to modify html handlers for scraping.
  - Improved form submission emulator to respect unnamed submit-type elements.

0.19 2016/05/23
  - Updated tests in favor of Mojo::Home recent changes

0.18 2016/03/16
  - Added cap attribute for queue to limit the length

0.17 2016/03/15
  - Fixed form submition emulation to work

0.16 2016/01/14
  - Added req event for such as request modification (gfdev).

0.15 2015/11/04
  - Fixed a bug where Mysql queue has not skipped correctly.
  - Fixed a bug where form submition emulation has failed on certain condition.

0.14 2015/10/30
  - Added experimental MySQL support for queue (harshals).
  - Improved documentations.
  - Added some example codes.
  - Improved form submittion emulation on unselected and multi selected options.

0.13 2015/03/25
  - Removed referrer_url attribute from job class.
  - Changed scrape API significantly.
  - Added context option for scrape method.
  - Improved documentations.
  - Improved examples.

0.12 2015/02/27
  - Updated dependency to Mojolicious v6.0.

0.11 2015/02/19
  - Updated dependency to Mojolicious v5.79.
  - Fixed a bug where connection count didn't detected for redirected urls.
  - Fixed a bug on checkbot example.
  - Fixed small bug on form submittion.
  - Fixed small bug on base tag detection.
  - Fixed small bug on charset detection.
  - Removed refer event in favor of callback for scraper.
  - Removed peeping server feature at all.
  - Deprecated resolved_uri attribute of Job class in favor of url.
  - Deprecated original_uri in favor of original_url.
  - Improved to crawl sitemap.xml too.
  - Improved html document detection by accepting more html like mime types.
  - Improved to use less memory.
  - Improved internal codes.
  - Added clock_speed attribute.

0.10 2015/02/08
  - Removed collect_urls_html method in favor of scrape.
  - Improved documentaion.
  - Improved url detection in CSSs.
  - Improved element handlers that is now well customizable with CSS selectors.

0.09 2015/02/03
  - Removed depth option.
  - Changed URL collecting method to instance-methods.
  - Improved form manipulation for emulating manual submition.
  - Improved documentation.

0.08 2015/01/29
  - Renamed browse method to scrape.
  - Improved documents.

0.07 2015/01/26
  - Removed additional_props on job class.
  - Renamed discover method to browse.
  - Removed peeping attribute in favor of peeping_port.
  - Improved error event API.
  - Added a feature for auto stop crawling when queue get empty.
  - Improved checkbot example.
  - Improved tests.
  - Improved documents.

0.06 2014/09/20
  - Improved URL detection.

0.05 2014/09/15
  - Fixed a bug on request body generation.

0.04 2014/09/07
  - Fixed class and method terminology.

0.03 2014/09/07
  - Improved that resolved URIs are alway a Mojo::URL instance.
  - Added original_uri to retrieve original uri from redirect history.
  - Added requeue method to re-try in case an error occured.

0.02 2014/09/07
  - Recovering failing release

0.01 2014/09/07
  - initial release


lib/Mojo/Crawler.pm
package Mojo::Crawler;
use strict;
use warnings;
use 5.010;
use Mojo::Base 'Mojo::EventEmitter';
use Mojo::Crawler::Queue;
use Mojo::Crawler::UserAgent;
use Mojo::Message::Request;
use Mojo::Util qw{md5_sum xml_escape dumper};
use List::Util;
our $VERSION = '0.01';

has active_conn => 0;
has 'crawler_loop_id';
has depth => 10;
has fix => sub { {} };
has active_conns_per_host => sub { {} };
has max_conn => 1;
has max_conn_per_host => 1;
has 'peeping';
has 'peeping_port';
has peeping_max_length => 30000;
has queues => sub { [] };
has 'ua' => sub { Mojo::Crawler::UserAgent->new };
has 'ua_name' => "mojo-crawler/$VERSION (+https://github.com/jamadam/mojo-crawler)";
has 'shuffle';

sub crawl {
    my ($self) = @_;
    
    $self->init;
    
    die 'No queue is given' if (! scalar @{$self->queues});
    
    $self->emit('start');
    
    Mojo::IOLoop->start unless Mojo::IOLoop->is_running;
}

sub init {
    my ($self) = @_;
    
    $self->on('empty', sub { say "Queue is drained out." })
                                        unless $self->has_subscribers('empty');
    $self->on('error', sub { say $_[1] })
                                        unless $self->has_subscribers('error');
    $self->on('res', sub { $_[1]->() })
                                        unless $self->has_subscribers('res');
    $self->on('refer', sub { $_[1]->() })
                                        unless $self->has_subscribers('refer');
    
    $self->ua->transactor->name($self->ua_name);
    $self->ua->max_redirects(5);
    
    my $loop_id = Mojo::IOLoop->recurring(0.25 => sub {
        $self->process_queue(@_);
    });
    
    $self->crawler_loop_id($loop_id);
    
    # notify queue is drained out
    Mojo::IOLoop->recurring(5 => sub {
        $self->emit('empty') if (! scalar @{$self->{queues}});
    });
    
    if ($self->peeping || $self->peeping_port) {
        # peeping API server
        my $id = Mojo::IOLoop->server({port => $self->peeping_port}, sub {
            $self->peeping_handler(@_);
        });
        $self->peeping_port(Mojo::IOLoop->acceptor($id)->handle->sockport);
    }
    
    if ($self->shuffle) {
        Mojo::IOLoop->recurring($self->shuffle => sub {
            @{$self->{queues}} = List::Util::shuffle @{$self->{queues}};
        });
    }
}

sub process_queue {
    my $self = shift;
    
    return unless ($self->{queues}->[0] &&
                $self->_mod_busyness($self->{queues}->[0]->resolved_uri, 1));
    
    my $queue = shift @{$self->{queues}};
    my $uri = $queue->resolved_uri;
    
    $self->ua->get($uri, sub {
        $self->_mod_busyness($uri, -1);
        
        my ($ua, $tx) = @_;
        
        $queue->redirect(urls_redirect($tx));
        
        my $res = $tx->res;
        
        if (!$res->code) {
            my $msg = ($res->error) ? $res->error->{message} : 'Unknown error';
            my $url = $queue->resolved_uri;
            $self->emit('error',
                        "An error occured during crawling $url: $msg", $queue);
            return;
        }
        
        $self->emit('res', sub { $self->discover($res, $queue) }, $queue, $res);
    });
}

sub urls_redirect {
    my $tx = shift;
    my @urls;
    @urls = urls_redirect($tx->previous) if ($tx->previous);
    unshift(@urls, $tx->req->url->userinfo(undef)->to_string);
    return @urls;
}

sub say_start {
    my $self = shift;
    
    print <<"EOF";
----------------------------------------
Crawling is starting with @{[ $self->queues->[0]->resolved_uri ]}
Max Connection  : @{[ $self->max_conn ]}
Depth           : @{[ $self->depth ]}
User Agent      : @{[ $self->ua_name ]}
EOF

    print <<"EOF" if ($self->peeping_port);
Peeping API is available at following URL
    http://127.0.0.1:@{[ $self->peeping_port ]}/
EOF
    
    print <<"EOF";
----------------------------------------
EOF
}

sub peeping_handler {
    my ($self, $loop, $stream) = @_;
    $stream->on(read => sub {
        my ($stream, $bytes) = @_;
        
        my $path = Mojo::Message::Request->new->parse($bytes)->url->path;
        
        if ($path =~ qr{^/queues}) {
            my $res = sprintf('%s Queues are left.', scalar @{$self->queues});
            $stream->write("HTTP/1.1 200 OK\n\n");
            $stream->write($res, sub {shift->close});
            return;
        }
        
        if ($path =~ qr{^/dumper/(\w+)} && defined $self->{$1}) {
            my $res = substr(dumper($self->{$1}), 0, $self->peeping_max_length);
            $stream->write("HTTP/1.1 200 OK\n\n");
            $stream->write($res, sub {shift->close});
            return;
        }
        
        $stream->write(
                    "HTTP/1.1 404 NOT FOUND\n\nNOT FOUND", sub {shift->close});
    });
}

sub discover {
    my ($self, $res, $queue) = @_;
    
    return if ($res->code != 200);
    return if ($self->depth && $queue->depth >= $self->depth);
    
    my $base = $queue->resolved_uri;
    my $type = $res->headers->content_type;
    
    if ($type && $type =~ qr{text/(html|xml)} &&
                                (my $base_tag = $res->dom->at('base'))) {
        $base = resolve_href($base, $base_tag->attr('href'));
    }
    
    my $cb = sub {
        my ($url, $dom) = @_;
        
        $url = Mojo::URL->new($url);
        
        return unless
                (!$url->scheme || $url->scheme =~ qr{http|https|ftp|ws|wss});
        
        my $child = $queue->child(
            resolved_uri => resolve_href($base, $url), literal_uri => $url);
        
        $self->emit('refer', sub {
            $self->enqueue($_[0] || $child);
        }, $child, $dom || $queue->resolved_uri);
    };
    
    if ($type && $type =~ qr{text/(html|xml)}) {
        my $encode = guess_encoding($res) || 'utf-8';
        my $body = Encode::decode($encode, $res->body);
        collect_urls_html(Mojo::DOM->new($body), $cb);
    }
    
    if ($type && $type =~ qr{text/css}) {
        my $encode  = guess_encoding_css($res) || 'utf-8';
        my $body    = Encode::decode($encode, $res->body);
        collect_urls_css($body, $cb);
    }
};

sub enqueue {
    my ($self, @queues) = @_;
    
    for (@queues) {
        $_ = Mojo::Crawler::Queue->new(resolved_uri => $_)
                        unless (ref $_ && ref $_ eq 'Mojo::Crawler::Queue');
        
        my $md5 = md5_sum($_->resolved_uri);
        
        if (!exists $self->fix->{$md5}) {
            $self->fix->{$md5} = undef;
            push(@{$self->{queues}}, $_);
        }
    }
}

sub collect_urls_html {
    my ($dom, $cb) = @_;
    
    $dom->find('script, link, a, img, area, embed, frame, iframe, input,
                                    meta[http\-equiv=Refresh]')->each(sub {
        my $dom = shift;
        if (my $href = $dom->{href} || $dom->{src} ||
            $dom->{content} && ($dom->{content} =~ qr{URL=(.+)}i)[0]) {
            $cb->($href, $dom);
        }
    });
    $dom->find('form')->each(sub {
        my $dom = shift;
        $cb->($dom->{action}, $dom) if ($dom->{action});
    });
    $dom->find('style')->each(sub {
        my $dom = shift;
        collect_urls_css($dom->content || '', sub { $cb->(shift, $dom) });
    });
    $dom->find('*[style]')->each(sub {
        my $dom = shift;
        collect_urls_css($dom->{style}, sub { $cb->(shift, $dom) });
    });
}

sub collect_urls_css {
    my ($str, $cb) = @_;
    $str =~ s{/\*.+?\*/}{}gs;
    my @urls = ($str =~ m{url\(['"]?(.+?)['"]?\)}g);
    $cb->($_) for (@urls);
}

my $charset_re = qr{\bcharset\s*=\s*['"]?([a-zA-Z0-9_\-]+)['"]?}i;

sub guess_encoding_css {
    my $res     = shift;
    my $type    = $res->headers->content_type;
    my $charset = ($type =~ $charset_re)[0];
    $charset =
        ($res->body =~ qr{^\s*\@charset ['"](.+?)['"];}is)[0] if (! $charset);
    return $charset;
}

sub guess_encoding {
    my $res     = shift;
    my $type    = $res->headers->content_type;
    my $charset = ($type =~ $charset_re)[0];
    if (! $charset && (my $head = ($res->body =~ qr{<head>(.+)</head>}is)[0])) {
        Mojo::DOM->new($head)->find('meta[http\-equiv=Content-Type]')->each(sub{
            $charset = (shift->{content} =~ $charset_re)[0];
        });
    }
    return $charset;
}

sub resolve_href {
    my ($base, $href) = @_;
    $href = ref $href ? $href : Mojo::URL->new($href);
    $base = ref $base ? $base : Mojo::URL->new($base);
    my $abs = $href->to_abs($base)->fragment(undef);
    while ($abs->path->parts->[0] && $abs->path->parts->[0] =~ /^\./) {
        shift @{$abs->path->parts};
    }
    $abs->path->trailing_slash($base->path->trailing_slash) if (!$href->path->to_string);
    return $abs;
}

sub _mod_busyness {
    my ($self, $uri, $inc) = @_;
    my $key = _host_key($uri);
    my $hosts = $self->active_conns_per_host;
    
    return if ($inc > 0 && ($self->active_conn >= $self->max_conn ||
                        ($hosts->{$key} || 0) >= $self->max_conn_per_host));
    
    $self->{active_conn} += $inc;
    $hosts->{$key} += $inc;
    delete($hosts->{$key}) unless ($hosts->{$key});
    return 1;
}

sub _host_key {
    my $uri = ref $_[0] ? $_[0] : Mojo::URL->new($_[0]);
    my $key = $uri->scheme. '://'. $uri->ihost;
    
    if (my $port = $uri->port) {
        if (($uri->scheme eq 'https' && $port != 443) ||
                                    ($uri->scheme eq 'http' && $port != 80)) {
            $key .= ':'. $port;
        }
    }
    
    return $key;
}

1;

=head1 NAME

Mojo::Crawler - A web crawling framework for Perl

=head1 SYNOPSIS

    use strict;
    use warnings;
    use utf8;
    use Mojo::Crawler;
    use 5.10.0;
    
    my $bot = Mojo::Crawler->new;
    my %count;
    
    $bot->on(res => sub {
        my ($bot, $discover, $queue, $res) = @_;
        
        $count{$res->code}++;
        
        if ($res->code == 404) {
            say sprintf('404 occured! : %s referred by %s',
                        $queue->resolved_uri, $queue->referrer->resolved_uri);
        }
        
        my @disp_seed;
        push(@disp_seed, sprintf('%s:%s', $_, $count{$_})) for (keys %count);
        
        $| = 1;
        print(join(' / ', @disp_seed), ' ' x 30);
        print("\r");
        
        $discover->();
    });
    
    $bot->on(refer => sub {
        my ($bot, $enqueue, $queue, $context) = @_;
        $enqueue->();
    });
    
    $bot->enqueue('http://example.com/');
    $bot->peeping_port(3001);
    $bot->crawl;

=head1 DESCRIPTION

Mojo-Crawler is a web crawling framework for Perl.

=head1 ATTRIBUTE

Mojo::Crawler inherits all attributes from Mojo::EventEmitter and implements the
following new ones.

=head2 active_conn

A number of current connections.

=head2 active_conns_per_host

A number of current connections per host.

=head2 depth

A number of max depth to crawl.

=head2 fix

=head2 active_conns_per_host

A hash contains host name for key and last requested timestamp for value.

=head2 max_conn

A number of Max connections.

=head2 max_conn_per_host

A number of Max connections per host.

=head2 peeping_max_length

Max length of peeping API content.

=head2 queues

FIFO array contains Mojo::Crawler::Queue objects.

=head1 EVENTS

Mojo::Crawler inherits all events from Mojo::EventEmitter and implements the
following new ones.

=head2 res

Emitted when crawler got response from server.

    $bot->on(res => sub {
        my ($bot, $discover, $queue, $res) = @_;
        if (...) {
            $discover->();
        } else {
            # DO NOTHING
        }
    });

=head2 refer

Emitted when new URI is found. You can enqueue new URIs conditionally with the callback.

    $bot->on(refer => sub {
        my ($bot, $enqueue, $queue, $context) = @_;
        if (...) {
            $enqueue->();
        } elseif (...) {
            $enqueue->(...); # maybe different url
        } else {
            # DO NOTHING
        }
    });

=head2 empty

Emitted when queue length got zero. The length is checked every 5 second.

    $bot->on(refer => sub {
        say "Queue is drained out.";
    });

=head2 error

Emitted when user agent returns no status code for request. Possibly causeed by
network errors or un-responsible servers.

    $bot->on(refer => sub {
        say "error: $_[1]";
    });

=head2 res

Emitted right before crawl is started.

    $bot->on(start => sub {
        my $self = shift;
        ...
    });

Note that server errors such as 404 or 500 cannot be catched with the event.
Consider res event for the use case instead of this.

=head1 METHODS

Mojo::Crawler inherits all methods from Mojo::EventEmitter and implements the
following new ones.

=head2 crawl

Start crawling loop.

=head2 init

Initialize crawler settings.

=head2 process_queue

Process a queue.

=head2 urls_redirect

Replace the resolved URI of queue and append redirect history into queue

=head2 say_start

Displays starting messages to STDOUT

=head2 peeping_handler

peeping API dispatcher.

=head2 discover

Parses and discovers lins in a web page.

=head2 enqueue

Append a queue with a URI or Mojo::Crawler::Queue object.

=head2 collect_urls_html

Collects URLs out of HTML.

=head2 collect_urls_css

Collects URLs out of CSS.

=head2 guess_encoding_css

Guesses encoding of CSS

=head2 guess_encoding

Guesses encoding of HTML

=head2 resolve_href

Resolves URLs with a base URL.

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) jamadam

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
